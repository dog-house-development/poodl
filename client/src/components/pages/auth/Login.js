import React, { Component } from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { loginUser } from "../../../actions/authActions";
import Form from "../../inputs/Form";

const propTypes = {
  loginUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired
};

class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: "",
      password: "",
      errors: {}
    };
  }

  componentDidMount() {
    // If logged in and user navigates to Login page, should redirect them to dashboard
    if (this.props.auth.isAuthenticated) {
      this.props.history.push("/dashboard");
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.auth.isAuthenticated) {
      this.props.history.push("/dashboard");
    }

    if (nextProps.errors) {
      this.setState({
        errors: nextProps.errors
      });
    }
  }

  onChange = e => {
    this.setState({ [e.target.id]: e.target.value });
  };

  onSubmit = e => {
    e.preventDefault();

    const userData = {
      email: this.state.email,
      password: this.state.password
    };

    this.props.loginUser(userData);
  };

  getFields = () => {
    const { errors } = this.state;
    const fields = [
      {
        id: "email",
        type: "email",
        label: "Email",
        value: this.state.email,
        onChange: this.onChange,
        error: errors.email,
        placeholder: "Email..."
      },
      {
        id: "password",
        type: "password",
        label: "Password",
        value: this.state.password,
        onChange: this.onChange,
        error: errors.email,
        placeholder: "Password......"
      }
    ];
    return fields;
  };

  render() {
    return (
      <div>
        <Link to="/">Back to home</Link>
        <div>
          <h4>
            <b>Login</b> below
          </h4>
          <p>
            Don't have an account? <Link to="/register">Register</Link>
          </p>
        </div>
        <Form
          fields={this.getFields()}
          onSubmit={this.onSubmit}
          noValidate
          buttonLabel="Log in"
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

Login.propTypes = propTypes;

export default connect(
  mapStateToProps,
  { loginUser }
)(Login);
