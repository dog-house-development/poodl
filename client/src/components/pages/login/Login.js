import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { Link, Redirect, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import _ from 'lodash';

import AuthActions from '../../../actions/authActions';
import Form from '../../ui/Form';

const propTypes = {
    auth: PropTypes.object.isRequired
};

export class Login extends Component {
    constructor() {
        super();
        this.state = {
            email: '',
            password: '',
            errors: {}
        };
    }

    componentDidMount() {
        // If logged in and admin navigates to Login page, should redirect them to dashboard
        if (this.props.auth.isAuthenticated) {
            this.props.history.push('/dashboard');
        }
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.errors) {
            this.setState({
                errors: nextProps.errors
            });
        }
    }

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = e => {
        e.preventDefault();

        const adminData = {
            email: this.state.email,
            password: this.state.password
        };

        this.props.authActions.loginAdmin(adminData);
    };

    getFields = () => {
        const { errors } = this.state;
        const fields = [
            {
                id: 'email',
                type: 'email',
                label: 'Email',
                value: this.state.email,
                onChange: this.onChange,
                error: errors.email,
                placeholder: 'Email...',
                autoComplete: 'on'
            },
            {
                id: 'password',
                type: 'password',
                label: 'Password',
                value: this.state.password,
                onChange: this.onChange,
                error: errors.password,
                placeholder: 'Password......',
                autoComplete: 'on'
            }
        ];
        return fields;
    };

    getFormErrors() {
        let errors = [];
        _.each(this.props.errors, (value, key) => {
            if (key.toLowerCase() === 'emailnotfound' || key.toLowerCase() === 'passwordincorrect') {
                errors.push('Email and password combination not found');
            } else if (key !== 'email' && key !== 'password') {
                errors.push(value);
            }
        });
        return errors;
    }

    render() {
        if (this.props.auth.isAuthenticated) {
            // take the saved previous attempted url and redirect to it otherwise dashboard
            const { from } = this.props.location.state || { from: { pathname: '/dashboard' } };
            return <Redirect to={from} />;
        }
        return (
            <div className="login-container">
                <Link to="/" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to home
                </Link>
                <Form
                    fields={this.getFields()}
                    onSubmit={this.onSubmit}
                    noValidate
                    buttonLabel="Log in"
                    formTitle="Log In"
                    errors={this.getFormErrors()}
                />
                <div className="center">
                    <p>
                        Forgot password?{' '}
                        <Link to="/login" className="button small secondary">
                            Reset
                        </Link>
                    </p>
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        errors: state.auth.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        authActions: bindActionCreators(AuthActions, dispatch)
    };
};

Login.propTypes = propTypes;

export default withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )(Login)
);
