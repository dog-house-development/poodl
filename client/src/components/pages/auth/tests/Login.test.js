import React from "react";
import ReactDOM from "react-dom";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import _ from "lodash";

import { Login, mapStateToProps, mapDispatchToProps } from "../Login";

configure({ adapter: new Adapter() });

describe("Login tests", () => {
  let wrapper, instance;
  const setInstanceAndWrapper = (_props = {}, _state = {}) => {
    const state = _.assign(
      {},
      {
        auth: {
          isAuthenticated: false,
          loading: false
        },
        errors: {}
      },
      _state
    );
    const props = _.assign({}, { history: ["/login"] }, _props);
    wrapper = shallow(
      <Login
        {..._.assign(
          {},
          props,
          mapStateToProps(state, props),
          mapDispatchToProps(jasmine.createSpy("dispatch"))
        )}
      />
    );
    instance = wrapper.instance();
    wrapper.setState({
      email: "test@test.test",
      password: "abc123",
      errors: {}
    });
  };

  beforeEach(() => {
    setInstanceAndWrapper();
  });

  describe("componentDidMount", () => {
    it("should redirect if user is authenticated", () => {
      const authenticatedState = {
        auth: {
          isAuthenticated: true,
          loading: false
        }
      };
      setInstanceAndWrapper({}, authenticatedState);
      const newInstanceProps = _.concat(instance.props.history, "/dashboard");
      instance.componentDidMount();
      expect(instance.props.history).toEqual(newInstanceProps);
    });
  });

  describe("render", () => {
    it("should render correctly", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
