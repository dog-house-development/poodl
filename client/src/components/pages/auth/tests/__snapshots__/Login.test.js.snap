// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests render should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    auth={
      Object {
        "isAuthenticated": false,
        "loading": false,
      }
    }
    errors={Object {}}
    history={
      Array [
        "/login",
      ]
    }
    loginAdmin={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          replace={false}
          to="/"
        >
          Back to home
        </Link>,
        <div>
          <h4>
            <b>
              Login
            </b>
             below
          </h4>
          <p>
            Don't have an account? 
            <Link
              replace={false}
              to="/register"
            >
              Register
            </Link>
          </p>
        </div>,
        <Form
          buttonLabel="Log in"
          fields={
            Array [
              Object {
                "error": undefined,
                "id": "email",
                "label": "Email",
                "onChange": [Function],
                "placeholder": "Email...",
                "type": "email",
                "value": "test@test.test",
              },
              Object {
                "error": undefined,
                "id": "password",
                "label": "Password",
                "onChange": [Function],
                "placeholder": "Password......",
                "type": "password",
                "value": "abc123",
              },
            ]
          }
          noValidate={true}
          onSubmit={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Back to home",
          "replace": false,
          "to": "/",
        },
        "ref": null,
        "rendered": "Back to home",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4>
              <b>
                Login
              </b>
               below
            </h4>,
            <p>
              Don't have an account? 
              <Link
                replace={false}
                to="/register"
              >
                Register
              </Link>
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <b>
                  Login
                </b>,
                " below",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Login",
                },
                "ref": null,
                "rendered": "Login",
                "type": "b",
              },
              " below",
            ],
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Don't have an account? ",
                <Link
                  replace={false}
                  to="/register"
                >
                  Register
                </Link>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Don't have an account? ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Register",
                  "replace": false,
                  "to": "/register",
                },
                "ref": null,
                "rendered": "Register",
                "type": [Function],
              },
            ],
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "buttonLabel": "Log in",
          "fields": Array [
            Object {
              "error": undefined,
              "id": "email",
              "label": "Email",
              "onChange": [Function],
              "placeholder": "Email...",
              "type": "email",
              "value": "test@test.test",
            },
            Object {
              "error": undefined,
              "id": "password",
              "label": "Password",
              "onChange": [Function],
              "placeholder": "Password......",
              "type": "password",
              "value": "abc123",
            },
          ],
          "noValidate": true,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            replace={false}
            to="/"
          >
            Back to home
          </Link>,
          <div>
            <h4>
              <b>
                Login
              </b>
               below
            </h4>
            <p>
              Don't have an account? 
              <Link
                replace={false}
                to="/register"
              >
                Register
              </Link>
            </p>
          </div>,
          <Form
            buttonLabel="Log in"
            fields={
              Array [
                Object {
                  "error": undefined,
                  "id": "email",
                  "label": "Email",
                  "onChange": [Function],
                  "placeholder": "Email...",
                  "type": "email",
                  "value": "test@test.test",
                },
                Object {
                  "error": undefined,
                  "id": "password",
                  "label": "Password",
                  "onChange": [Function],
                  "placeholder": "Password......",
                  "type": "password",
                  "value": "abc123",
                },
              ]
            }
            noValidate={true}
            onSubmit={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Back to home",
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": "Back to home",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h4>
                <b>
                  Login
                </b>
                 below
              </h4>,
              <p>
                Don't have an account? 
                <Link
                  replace={false}
                  to="/register"
                >
                  Register
                </Link>
              </p>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <b>
                    Login
                  </b>,
                  " below",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Login",
                  },
                  "ref": null,
                  "rendered": "Login",
                  "type": "b",
                },
                " below",
              ],
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Don't have an account? ",
                  <Link
                    replace={false}
                    to="/register"
                  >
                    Register
                  </Link>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Don't have an account? ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Register",
                    "replace": false,
                    "to": "/register",
                  },
                  "ref": null,
                  "rendered": "Register",
                  "type": [Function],
                },
              ],
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "buttonLabel": "Log in",
            "fields": Array [
              Object {
                "error": undefined,
                "id": "email",
                "label": "Email",
                "onChange": [Function],
                "placeholder": "Email...",
                "type": "email",
                "value": "test@test.test",
              },
              Object {
                "error": undefined,
                "id": "password",
                "label": "Password",
                "onChange": [Function],
                "placeholder": "Password......",
                "type": "password",
                "value": "abc123",
              },
            ],
            "noValidate": true,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
